### PR TITLE
Add support for digest auth

### PR DESCRIPTION
**Pull request by eliangcs**
Adds support for HTTP Digest Auth. Example app https://github.com/zapier/zapier-platform-example-app-digest-auth demonstrates the usage with tests.

### CODE REVIEW

File: src/http-middlewares/before/add-digest-auth-header.js [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+
+const crypto = require('crypto');
+const urllib = require('url');
+
+const fetch = require('node-fetch');
+
+const { NotImplementedError } = require('../../errors');
+
+// This function splits a comma-separated string described by RFC 2068 Section 2.
+// Ported from https://github.com/python/cpython/blob/f081fd83/Lib/urllib/request.py#L1399-L1440
```
Comment (xavdid) [inline src/http-middlewares/before/add-digest-auth-header.js:11]: Most minor of comments, but I'd pin this to a commit so it still makes sense if they update that file and the function moves around: `https://github.com/python/cpython/blob/f081fd83032be48aefdb1bbcc38ab5deb03785d5/Lib/urllib/request.py#L1399-L1440`

File: src/http-middlewares/before/add-digest-auth-header.js [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+};
+
+// Parse lists of key, value pairs as described by RFC 2068 Section 2 and convert them
+// into a python dict.
+// Ported from https://github.com/requests/requests/blob/d2962f1d/requests/utils.py#L342-L373
+const parseDictHeader = s => {
+  const res = {};
+
+  const items = parseHttpList(s);
+  for (let item of items) {
```
Comment (xavdid) [inline src/http-middlewares/before/add-digest-auth-header.js]: I'd change to `forEach` as not to block the event loop

File: src/http-middlewares/before/add-digest-auth-header.js [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+    if (item.indexOf('=') === -1) {
```
Comment (xavdid) [inline src/http-middlewares/before/add-digest-auth-header.js]: I find `!item.includes('=')` cleaner, but I won't fight about it

**Review by xavdid — APPROVED** [review]
a couple of minor things, but otherwise it looks great! thanks a ton!

File: src/http-middlewares/before/add-digest-auth-header.js (hunk 1 above)
Comment (eliangcs) [reply src/http-middlewares/before/add-digest-auth-header.js:11]: It's already a commit-specific URL. I shortened the commit hash just to make the URL shorter.

File: src/http-middlewares/before/add-digest-auth-header.js (hunk 3 above)
Comment (eliangcs) [reply src/http-middlewares/before/add-digest-auth-header.js]: Fixed in 4f15e89. Thanks!

File: src/http-middlewares/before/add-digest-auth-header.js (hunk 2 above)
Comment (eliangcs) [reply src/http-middlewares/before/add-digest-auth-header.js]: Do you mean `forEach` can be asynchronous? I looked it up, but I can't find any resources suggesting `forEach` being able not to block the event loop. But I fixed it anyway in 4f15e89 just to keep the coding style consistent with elsewhere.

File: src/http-middlewares/before/add-digest-auth-header.js (hunk 1 above)
Comment (xavdid) [reply src/http-middlewares/before/add-digest-auth-header.js:11]: ah you're very correct!

File: src/http-middlewares/before/add-digest-auth-header.js (hunk 2 above)
Comment (xavdid) [reply src/http-middlewares/before/add-digest-auth-header.js]: I can't find the comment now, but @stevelikesmusic (i think?) had mentioned something about the difference mattering. consistency is good too though!

**Comment by xavdid** [comment]
👍

File: src/http-middlewares/before/add-digest-auth-header.js (hunk 2 above)
Comment (stevelikesmusic) [reply src/http-middlewares/before/add-digest-auth-header.js]: I'm not sure I recall the comment 🤔 

It looks like all of this is synchronous besides the one `fetch`. The only way I know to process these in parallel would be to create `Promises` and then process those with something like `await Promise.all(items.map(asynchronousProcess))`. But that seems like an optimization we don't yet need here? I may be missing some context as well.

Other than what `for..of` (any iterable) and `forEach` (arrays only) can operate on, I think they are pretty much the same in how they execute.

Comment (xavdid) [reply src/http-middlewares/before/add-digest-auth-header.js]: ¯\\\_(ツ)_/¯

**Pull request merged**